crates/env/src/lib.rs

```diff
@@ -163,3 +163,16 @@ cfg_if::cfg_if! {
         }
     }
 }
+
+cfg_if::cfg_if! {
+    if #[cfg(feature = "std")] {
+        #[macro_export]
+        /// Prepend contract message call with value transfer. Used for tests in off-chain environment.
+        macro_rules! pay_with_call {
+            ($con:ident . $msg:ident ( $($params:ty)? ) , $amt:expr) => {{
+                transfer_in::<Environment>($amt);
+                $con.$msg($($params:ty)?)
+            }}
+        }
+    }
+}
```

Comment:
Can we replace this with a function that accepts a closure instead?

Something like:

```rust
pub fn pay_with_call_fn<F, O, T>(f: F, amount: T::Balance) -> O
where
    F: FnOnce() -> O,
    T: Environment<Balance = u128>,
{
    transfer_in::<T>(amount);
    f()
}
```

This way we can make it more easily discoverable with the rest of the testing functions

Reply:
I was thinking to impl it this way at first. But how to make it in the way so that it can work with messages returning something? I want to cover all possible message signatures, the ones taking parameters and the ones returning any value. I'm not sure if I get how your example solves this.

Reply:
Hmm alright fair enough, let's stick with the macro